security/src/main/java/io/airlift/security/client/ExpiringAuthenticationStore.java

```diff
@@ -0,0 +1,90 @@
+package io.airlift.security.client;
+
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+import io.airlift.units.Duration;
+import org.eclipse.jetty.client.api.Authentication;
+import org.eclipse.jetty.client.api.AuthenticationStore;
+
+import java.net.URI;
+import java.util.concurrent.TimeUnit;
+
+import static java.util.Objects.requireNonNull;
+
+public class ExpiringAuthenticationStore
+        implements AuthenticationStore
+{
+    private static final int CACHE_SIZE = 10000;
+    private static final Duration CACHE_EXPIRE_TIME = new Duration(5, TimeUnit.MINUTES);
+    private static final int CONCURRENCY_LEVEL = 16;
+    private final Cache<URI, Authentication.Result> results;
+
+    // Only support SPNEGO authentication for now
+    private final SpnegoAuthentication authentication;
+
+    public ExpiringAuthenticationStore(SpnegoAuthentication authentication)
+    {
+        requireNonNull(authentication, "authentication is null");
+        this.authentication = authentication;
+        results = CacheBuilder.newBuilder()
+                .concurrencyLevel(CONCURRENCY_LEVEL)
+                .maximumSize(CACHE_SIZE)
+                .expireAfterWrite(CACHE_EXPIRE_TIME.roundTo(TimeUnit.MINUTES), TimeUnit.MINUTES).build();
+    }
+
+    @Override
+    public void addAuthentication(Authentication authentication)
+    {
+        throw new UnsupportedOperationException("addAuthentication is not supported");
+    }
+
+    @Override
+    public void removeAuthentication(Authentication authentication)
+    {
+        throw new UnsupportedOperationException("removeAuthentication is not supported");
+    }
+
+    @Override
+    public void clearAuthentications()
+    {
+        throw new UnsupportedOperationException("clearAuthentications is not supported");
+    }
+
+    @Override
+    public Authentication findAuthentication(String type, URI uri, String realm)
+    {
+        if (authentication.matches(type, uri, realm)) {
+            return authentication;
+        }
+        throw new UnsupportedOperationException(String.format("Authentication type %s is not supported", type));
```

Comment:
I think this should be `IllegalArgumentException`. `UnsupportedOperationException` usually means that the entire operation (this method call) is not supported. In this case it is dependent on the argument.